hadoop-hdds/docs/content/design/decommissioning.md

```diff
@@ -0,0 +1,720 @@
+
+The transient failure can violate the availability guarantees of Ozone; Since the user is telling us not to take corrective actions. Many times, the user does not understand the impact on availability while asking Ozone to ignore the failure.
+
+So this paper proposes the following definitions for Decommission and Maintenance of data nodes.
+
+__Decommission__ of a data node is deemed to be complete when SCM/Ozone completes the replica of all containers on decommissioned data node to other data nodes.That is, the expected count matches the healthy count of containers in the cluster.
+
+__Maintenance mode__ of a data node is complete if Ozone can guarantee at least one copy of every container is available in other healthy data nodes.
+
+## Examples 
+
+Here are some illustrative examples:
+
+1.  Let us say we have a container, which has only one copy and resides on Machine A. If the user wants to put machine A into maintenance mode; Ozone will make a replica before entering the maintenance mode.
+
+2. Suppose a container has two copies, and the user wants to put Machine A to maintenance mode. In this case; the Ozone understands that availability of the container is not affected and hence can decide to forgo replication.
+
+3. Suppose a container has two copies, and the user wants to put Machine A into maintenance mode. However, the user wants to put the machine into maintenance mode for one month. As the period of maintenance mode increases, the probability of data loss increases; hence, Ozone might choose to make a replica of the container even if we are entering maintenance mode.
+
+4. The semantics of decommissioning means that as long as we can find copies of containers in other machines, we can technically get away with calling decommission complete. Hence this clarification node; in the ordinary course of action; each decommission will create a replication flow for each container we have; however, it is possible to complete a decommission of a data node, even if we get a failure of the  data node being decommissioned. As long as we can find the other datanodes to replicate from and get the number of replicas needed backup to expected count we are good.
+
+5. Let us say we have a copy of a container replica on Machine A, B, and C. It is possible to decommission all three machines at the same time, as decommissioning is just a status indicator of the data node and until we finish the decommissioning process.
+
+
+The user-visible features for both of these  are very similar:
+
+Both Decommission and Maintenance mode can be canceled any time before the operation is marked as completed by SCM.
+
+Decommissioned nodes, if and when added back, shall be treated as new data nodes; if they have blocks or containers on them, they can be used to reconstruct data.
+
```

Comment:
Is it worth adding a requirement / note, that a node which is DECOMMISSIONED or IN MAINTENANCE can be used as a last resort for reads if it is heartbeating? HDFS has this feature, but I believe a node in the above states will only be used for reads of there are no other replicas available.

Reply:
Good question. I think there are two questions:

 1 Can the decommissioning node be used for read by the clients?
 *2Can the decommissioning node can be used as a source of replication?

AFAIK the current implementation of the api calls and pipeline management can support 1 easily: client can try to use decommissioning nodes and retry to read from other nodes if it's not available any more.

The second one is a harder question (for me). I would prefer to use healthy nodes as a replication source but in this case the replication can be limited.

Reply:
I guess we need to consider if the container choosing logic should see "DECOMMISSIONING" as a healthy node for replication, but it complicates things slightly as we should only use it as a last resort for reads - ie the other nodes have gone unavailable for some reason.

The decommissioning node is a also a great source for replication reads. Its guaranteed to be under no write load and if readers normally don't touch it (except as a last resort), it will be basically idle unless its used as a replication source.